src/widget/wsearchlineedit.cpp

```diff
@@ -324,60 +330,77 @@ QString WSearchLineEdit::getSearchText() const {
-                // no matter if it's a saved and unsaved query
-                if (findCurrentTextIndex() == 0 ||
-                        (findCurrentTextIndex() == -1 && !currentText().isEmpty())) {
-                    slotClearSearch();
-                    return true;
-                }
-            } else if (keyEvent->key() == Qt::Key_Down) {
-                // after clearing the text field the down key is expected to
-                // show the latest entry
-                if (currentText().isEmpty()) {
-                    setCurrentIndex(0);
-                    return true;
-                }
-                // in case the user entered a new search query
-                // and presses the down key, save the query for later recall
-                if (findCurrentTextIndex() == -1) {
-                    slotSaveSearch();
-                }
-            }
-        } else {
-            if (keyEvent->key() == Qt::Key_Backspace ||
-                    keyEvent->key() == Qt::Key_Delete) {
-                // remove the highlighted item from the list
-                deleteSelectedListItem();
-                return true;
-            }
-        }
-        if (keyEvent->key() == Qt::Key_Enter) {
-            if (findCurrentTextIndex() == -1) {
-                slotSaveSearch();
-            }
-            // The default handler will add the entry to the list,
-            // this already happened in slotSaveSearch
-            slotTriggerSearch();
-            return true;
-        } else if (keyEvent->key() == Qt::Key_Space &&
-                keyEvent->modifiers() == Qt::ControlModifier) {
-            // open/close popup on ctrl + space
-            if (view()->isVisible()) {
-                hidePopup();
-            } else {
-                showPopup();
-            }
+        const int key = keyEvent->key();
+        // Esc has already closed the popup by now and we don't want to process it.
+        // We don't need to handle Up/Down in the popup either.
+        // Any other keypress is forwarded.
+        if (key != Qt::Key_Escape &&
+                key != Qt::Key_Down &&
+                key != Qt::Key_Up) {
+            keyPressEvent(keyEvent);
             return true;
         }
-        // if the line edit has focus Ctrl + F selects the text
     }
     return QComboBox::eventFilter(obj, event);
 }
 
+void WSearchLineEdit::keyPressEvent(QKeyEvent* keyEvent) {
+    const int key = keyEvent->key();
+    if (view()->isVisible()) {
+        if (key == Qt::Key_Backspace ||
+                key == Qt::Key_Delete) {
+            // remove the highlighted item from the list
+            deleteSelectedListItem();
+            return;
+        }
+    }
+
+    if (key == Qt::Key_Up) {
+        // If we're at the top of the list the Up key clears the search bar,
+        // no matter if it's a saved or unsaved query.
+        // Otherwise Up is handled by the combobox itself.
+        if (findCurrentTextIndex() == 0 ||
```

Comment:
seems to me like `findCurrentTextIndex()` is not super cheap but side-effect free. It would make sense to introduce a local variable for it in this if arm.

Reply:
I didn't change the key cases, I only moved them from the eventFilter.
Thanks for the hint anyway, added in a fixup.